scipy/stats/_unuran/unuran_wrapper.pyx.templ

```diff
@@ -0,0 +1,1122 @@
+            _lock.release()
+            release_unuran_callback(&callback)
+
+    def ppf_hat(self, u):
+        """
+        Evaluate the inverse of the CDF of the hat distribution at `u`.
+
+        This call does not work for variant IA (immediate acceptance).
+        In this case the hat CDF is evaluated as if variant PS is used.
+
+        Parameters
+        ----------
+        u : array_like
+            An array of percentiles
+
+        Returns
+        -------
+        ppf_hat : array_like
+            Array of quantiles corresponding to the given percentiles.
+
+        Examples
+        --------
+        >>> from scipy.stats import TransformedDensityRejection
+        >>> from scipy.stats import norm
+        >>> from math import exp
+        >>> 
+        >>> class MyDist:
+        ...     def pdf(self, x):
+        ...         return exp(-0.5 * x**2)
+        ...     def dpdf(self, x):
+        ...         return -x * exp(-0.5 * x**2)
+        ... 
+        >>> dist = MyDist()
+        >>> rng = TransformedDensityRejection(dist)
+        >>> 
+        >>> rng.ppf_hat(0.5)
+        -0.00018050266342393984
+        >>> norm.ppf(0.5)
+        0.0
+        >>> u = np.linspace(0, 1, num=1000)
+        >>> ppf_hat = rng.ppf_hat(u)
+        """
+        u = np.asarray(u, dtype='d')
+        oshape = u.shape
+        u = u.ravel()
+        # UNU.RAN fills in ends of the support when u < 0 or u > 1 while
+        # SciPy fills in nans. Prefer SciPy behaviour.
+        cond0 = 0 <= u
+        cond1 = u <= 1
+        cond2 = cond0 & cond1
+        goodu = argsreduce(cond2, u)[0]
+        out = np.empty_like(u)
+        cdef double[::1] u_view = np.ascontiguousarray(goodu)
+        cdef double[::1] goodout = np.empty_like(u_view)
+        if cond2.any():
+            self._ppf_hat(&u_view[0], &goodout[0], len(goodu))
+        np.place(out, cond2, goodout)
+        np.place(out, ~cond2, np.nan)
+        return np.asarray(out).reshape(oshape)[()]
+
+
+cdef class DiscreteAliasUrn(Method):
+    r"""
+    Discrete Alias-Urn Method.
+
+    This method is used to sample from univariate discrete distributions with
+    a finite domain. It uses the probability vector of size :math:`N` or a
+    probability mass function with a finite support to generator random
+    numbers from the distribution.
+
+    Parameters
+    ----------
+    pv : array_like, optional
+        Probability vector (PV) of the distribution. If PV isn't available,
+        PMF of the distribution is expected.
+    dist : object, optional
+        An instance of a class with a 'pmf' method. Only required if PV isn't
+        available. The signature of the PMF is expected to be:
+        ``def pmf(k: int, *params) -> float``. i.e it should accecpt a Python
+        integer and return a Python float.
+    params : tuple, optional
+        Parameters that the PMF takes. If a probability vector is available,
+        this parameter is ignored. Default is an empty tuple.
```

Comment:
Whatever decision is made about [this](https://github.com/scipy/scipy/pull/14215/files#r684450336) would apply here, too.

Reply:
same as above: params has been removed